src/tools/cephfs/DataScan.cc

```diff
@@ -0,0 +1,1267 @@
+      "one node before running 'recover'" << std::endl;
+    return -EIO;
+  }
+
+  for (; i != i_end; ++i) {
+    const std::string oid = i->get_oid();
+#ifdef SHARDED_PGLS
+    if (i.get_progress() != progress) {
+      if (int(i.get_progress() * 100) / 5 != int(progress * 100) / 5) {
+        std::cerr << percentify(i.get_progress()) << "%" << std::endl;
+      }
+      progress = i.get_progress();
+    }
+#endif
+
+    uint64_t obj_name_ino = 0;
+    uint64_t obj_name_offset = 0;
+    r = parse_oid(oid, &obj_name_ino, &obj_name_offset);
+    if (r != 0) {
+      dout(4) << "Bad object name '" << oid << "', skipping" << dendl;
+      continue;
+    }
+
+    // We are only interested in 0th objects during this phase: we touched
+    // the other objects during recover_extents
+    if (obj_name_offset != 0) {
+      continue;
+    }
+
+    // Read backtrace
+    bool have_backtrace = false;
+    bufferlist parent_bl;
+    int r = data_io.getxattr(oid, "parent", parent_bl);
+    if (r == -ENODATA) {
+      dout(10) << "No backtrace on '" << oid << "'" << dendl;
+    } else if (r < 0) {
+      // TODO: accumulate these errors in a structure we can output
+      dout(4) << "Unexpected error on '" << oid << "': " << cpp_strerror(r) << dendl;
+      continue;
+    }
+
+    // Deserialize backtrace
+    inode_backtrace_t backtrace;
+    if (parent_bl.length()) {
+      try {
+        bufferlist::iterator q = parent_bl.begin();
+        backtrace.decode(q);
+        have_backtrace = true;
+      } catch (buffer::error &e) {
+        dout(4) << "Corrupt backtrace on '" << oid << "': " << e << dendl;
+      }
+    }
+
+    // Santity checking backtrace ino against object name
+    if (have_backtrace && backtrace.ino != obj_name_ino) {
+      dout(4) << "Backtrace ino 0x" << std::hex << backtrace.ino
+        << " doesn't match object name ino 0x" << obj_name_ino
+        << std::dec << dendl;
+      have_backtrace = false;
+    }
+
+    // Read size
+    uint64_t obj_size = 0;
+    time_t obj_mtime = 0;
+    r = data_io.stat(oid, &obj_size, &obj_mtime);
+    if (r != 0) {
+      dout(4) << "Cannot stat '" << oid << "': skipping" << dendl;
+      continue;
+    }
+
+    // Read accumulated size from scan_extents phase
+    uint64_t file_size = 0;
+    uint32_t chunk_size = g_default_file_layout.fl_object_size;
+    bufferlist scan_size_bl;
+    r = data_io.getxattr(oid, "scan_size", scan_size_bl);
+    if (r >= 0) {
+      try {
+        bufferlist::iterator scan_size_bl_iter = scan_size_bl.begin();
+        uint64_t max_obj_id;
+        uint64_t max_obj_size;
+        ::decode(max_obj_id, scan_size_bl_iter);
+        ::decode(max_obj_size, scan_size_bl_iter);
+
+        if (max_obj_id > 0) {
+          // If there are more objects, and this object is a power of two
+          // that is greater than the size of the last object, then make
+          // an educated guess that this object's size is probably the chunk
+          // size.
```

Comment:
This doesn't work for sparse files — I like the idea of gathering up the largest object better!